lib/oneboxer.rb

```diff
@@ -313,27 +313,82 @@ def self.allowed_iframe_origins
+                                ignore_redirects: ignore_redirects,
+                                ignore_hostnames: blocked_domains,
+                                force_get_hosts: force_get_hosts,
+                                force_custom_user_agent_hosts: force_custom_user_agent_hosts,
+                                preserve_fragment_url_hosts: preserve_fragment_url_hosts)
       uri = fd.resolve
+
+      if fd.status != :resolved
+        args = { link: url }
+        if fd.status == :invalid_address
+          args[:error_message] = I18n.t("errors.onebox.invalid_address", hostname: fd.hostname)
+        elsif fd.status_code
+          args[:error_message] = I18n.t("errors.onebox.error_response", status_code: fd.status_code)
+        end
+
+        error_box = blank_onebox
+        error_box[:preview] = preview_error_onebox(args)
+        return error_box
+      end
+
       return blank_onebox if uri.blank? || blocked_domains.map { |hostname| uri.hostname.match?(hostname) }.any?
 
       options = {
         max_width: 695,
         sanitize_config: Onebox::DiscourseOneboxSanitizeConfig::Config::DISCOURSE_ONEBOX,
         allowed_iframe_origins: allowed_iframe_origins,
         hostname: GlobalSetting.hostname,
+        facebook_app_access_token: SiteSetting.facebook_app_access_token,
       }
 
       options[:cookie] = fd.cookie if fd.cookie
 
       r = Onebox.preview(uri.to_s, options)
+      result = { onebox: r.to_s, preview: r&.placeholder_html.to_s }
+
+      # NOTE: Call r.errors after calling placeholder_html
+      if r.errors.any?
+        # TODO: verify this will work for all locales
+        connector = " #{I18n.t('errors.onebox.missing_data_word_connector')} "
+        error_msg = r.errors.keys.map(&:to_s).to_sentence(two_words_connector: connector, last_word_connector: connector)
+        error_message = I18n.t("errors.onebox.missing_data", missing_data_fields: error_msg)
+        args = r.data.merge(error_message: error_message)
+
+        if result[:preview].blank?
+          result[:preview] = preview_error_onebox(args)
+        else
+          doc = Nokogiri::HTML5::fragment(result[:preview])
+          aside = doc.at('aside')
+
+          if aside
+            # Add an error message to the preview that was returned
+            error_fragment = preview_error_onebox_fragment(args)
+            aside.add_child(error_fragment)
+            result[:preview] = doc.to_html
+          end
+        end
+      end
 
-      { onebox: r.to_s, preview: r&.placeholder_html.to_s }
+      result
     end
   end
 
+  def self.preview_error_onebox(args, is_fragment = false)
+    args[:title] ||= args[:link] if args[:link]
+    args[:error_message] = PrettyText.unescape_emoji(args[:error_message]) if args[:error_message]
+
+    if is_fragment
+      template = File.read("#{Rails.root}/lib/onebox/templates/preview_error_fragment_onebox.mustache")
```

Comment:
Are these read from the filesystem every time there's an error to render? Are there performance considerations here if the error was being generated many times?

Reply:
I've made a change. Is this the sort of thing you had in mind?

https://github.com/discourse/discourse/pull/11253/files#diff-4c68ae937b450c4abe514b2105ed4d8e138b0630de87c85cebc0cab3f9baabfbR389

Reply:
Yup, already approved, thanks.